### DISCUSSION TITLE
not working as intended: How to securely host fari "online" (i.e. exposed with port forwarding) - Linux

### DISCUSSION THREAD

**Discussion by XaverStiensmeier** [fariapp/fari-app#549] [Show and tell]
# How To Host Fari - Hands On
Comment: while this tutorial focuses on Linux, it should work for windows alike - installation processes differ though.

Make sure to read the [Caddy](#caddy) part. It's important for security. This is a highly simplified tutorial which will not *explain* every step in great detail, but explains how to **follow** each step in sufficient detail. That basically means: You might not understand every step, but if you follow them, everything will work. I would recommend to read about everything mentioned here as you are putting your system at risk whenever you follow a tutorial blindly.

## Download / Clone Fari
You find the code here [https://github.com/farirpgs/fari-app](https://github.com/farirpgs/fari-app). Just click on Code>Download Zip and unpack it into a nice folder on your computer - or clone it if you know how that works - both ways are equally possible.

### NVM
You can use NVM in order to manage multiple node installations. This is **not required**:
- [https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating](https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating) (installing nvm; restart terminal afterwards or execute the export)
- `nvm install node` (installs latest version)
- `node -v` to see if the installation was successful

## Host Fari Locally
We will now host Fari locally. That means no one from outside can access your instance. In the **Fari directory** execute:
```sh
npm run start
```
## Host Fari Exposed
In the Fari directory execute:
```sh
npm run start -- --host
```
If you do not have crazy router settings, still no one can access your instance over the internet, because you still need to forward your port. However, I highly recommend you to use caddy in between in order to have your instance password protected and not openly reachable by everyone.

#### Install Caddy
[https://caddyserver.com/docs/install](https://caddyserver.com/docs/install)
#### Creating Caddy Config
Caddy allows you to set an authentication process infront of your fari hosting. Once you enter the correct user - password pair, you get access to the fari instance. Otherwise, everybody with your ip and port could connect. Therefore, caddy is a **must have** in my opinion.

In order to use caddy, you need to create a caddy file. In that file under basicauth you need to set a username and enter a password hash. In order to hash your password, use `caddy hash-password` and copy the result into the caddyfile ([some words on hashing](https://security.stackexchange.com/questions/41447/why-is-password-hashing-considered-so-important) not relevant for this hands-on)
```
# caddyfile
:3000 {
    reverse_proxy localhost:1234

    basicauth / {
        <username> <password hash>
    }
}
```

#### Check If Caddy Is Already Running
```sh
sudo systemctl status caddy # if active
sudo systemctl stop caddy
```

You can also run caddy as a service, but for now I will assume that you don't want to run caddy all the time, but just when you actually run the session.

#### Run caddy
```sh
caddy run --config /path/to/your/caddyfile --adapter caddyfile
```

### Port Forwarding
Port forwarding is needed to allow other computers from outside of your network to reach your computer. Port forwarding is created/allowed/established on your router. Port forwarding is router specific best search "port forwarding \<router name\>" in oder to get specific information on your router.

If you don't port forward, you will not be able to be reached from outside. You need to forward the port 3000 if you use caddy or 1234 if not.

#### FritzBox Example
![PermitAccess](https://github.com/user-attachments/assets/934fc4b5-bf62-42d4-a290-0ad95e530262)
![Device](https://github.com/user-attachments/assets/86d48a14-78da-4ec1-bd1a-6eef5733bcf5)
![NewSharing](https://github.com/user-attachments/assets/a33f6d68-c5b5-4e35-92ce-37da2ef6bc80)
![NewSharing2](https://github.com/user-attachments/assets/5efdb743-6bdb-4059-beb8-e6df5f07f82d)

### How Can People Connect
Given that you have started Fari Exposed, Caddy and have created the port forwarding, people can now connect to your fari instance by entering into the browser:
```
<public-ip>:3000
```
If you don't know your public IP, visit [https://www.whatismyip.com/](https://www.whatismyip.com/) or another site like that. Afterwards, if you do run Caddy which I highly recommend, you can enter your password and access the fari instance. Test it with your phone, too, just to make 100% sure everything works.

Great! You did it!

## Common Issues
- Is caddy running? Check `localhost:3000`!
- Is fari running? Check `localhost:1234`!
- Caddy's password not working? Did you hash it in caddy?
- Did you forget to save the port forwarding.
- Did you forward the right port?
- Did your (public) IP Adress change in the mean time? You can let your router assign static private ips for your private network. You can ask your provider to provide you with a static public ip - I would not recommend that for safety reasons.
- You or your friend might not have [https://www.techopedia.com/definition/19025/dual-stack-network](https://www.techopedia.com/definition/19025/dual-stack-network). See if both ips are shown on [https://www.whatismyip.com/](https://www.whatismyip.com/).

Good luck!

[fari.pdf](https://github.com/user-attachments/files/16650087/fari.pdf)

**Comment by lynn0702** [comment]
This is neat, but may not be achieving the goals you were hoping for, assuming those goals were lessened hosting costs to fari.app or improved data privacy for your games. 

Be mindful that this is still consuming fari's liveblocks service hosting resources. This service enables Fari better managed connectivity for hosted games and their clients, i.e. you hosting a game for your friends.

Fari was not storing any of your game data, nor did it have any visibility into that data beyond connectivity analytics. Now those analytics are impacted because self-hosted instance calls to liveblocks aren't attributable to fari.app.

If your primary purpose is creating a sandbox to explore the app and test out local modifications then forge ahead.

**Reply by XaverStiensmeier** [reply]
Thank you for the reply! I have not looked deeply into the code, so I might have done something silly here. To make sure: My **intend** is to provide a way to run fari for users in case that the official fari page shuts down for some reason in the future. It's also pretty analogue to Foundry self-hosting which makes it feasible for regular VTT users, I think.

What are liveblock service hosting resources? How is the proposed solution still consuming fari's liveblocks service hosting resources? I can run that application without internet locally in exposed mode - so at least in theory it should work in a similar fashion when I use port forwarding over the internet. Of course it might be the case that additional features (and resources) are being used once connectivity is established - is that what you are implying? In that case providing a startup parameter that disables those optional services is probably the best solution to handle this. Sorry, if I misunderstood something.

EDIT: I have tested a bit more. I think I understand what you mean now. For "matchmaking" you use an externals service which is not included in the fari app, right? That service is paid for by fari official and once fari shuts down, that service would shut down, too. That's a bummer and renders my solution useless, I fear. What do you think? Maybe it is possible to "short" the matchmaking somehow...

**Reply by RPDeshaies** [reply]
Based on your edit, you are correct. Fari App is 100% a client side app without any server. Once bundled it's just front-end code and the server is just a normal http server that renders html, css and javascript files all on the client.

Except for live games. We rely on Liveblocks (and Liveblocks offers a generous open source plan to us) for this.

Even if you run Fari App somewhere else, it's still 100% a client side app and still relies on Liveblocks. 
If we run out of Liveblocks credits, online will stop working for you too.

That being said, Fari App isn't shutting down. It's just not receiving major upgrades and as long as Liveblocks gives us our plan, online games will continue to work.

Hope this clear things up!

**Reply by XaverStiensmeier** [reply]
I feel like these are two different topics, but maybe its still the right place for both:
I) So if I understand correctly, liveblock credits are given every time period and not like one large batch that is used until "live games" run out of juice? If that's the case, I feel like some people spread false rumours. Good to know! It sounded more like a "at point X in time we will have no credits left and its over". 
II) what is it that liveblock actually does? Is it just a data storage where both client sides can read and store data?
III) do you feel like it is feasible to replace that part of liveblock for such small setups?

I'm a bit sad that my idea isn't working as intended, not gonna lie, but that doesn't detract from your helpful explanation and work!
